### PR TITLE
xdg-shell: completely rewrite xdg-positioner

### PR DESCRIPTION
**Pull request by outfoxxed**
#### Describe your PR, what does it fix/add?
Completely replaces xdg-positioner's positioning implementation with one that actually works correctly, derived from the [one I wrote for quickshell a couple days ago](https://git.outfoxxed.me/outfoxxed/quickshell/src/commit/18563b1273166f6e5795e892642fd11f6b66e69a/src/core/popupanchor.cpp#L140).

This implementation has been tested and all cases work as expected.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
No

#### Is it ready for merging, or does it need work?
Needs https://github.com/hyprwm/hyprutils/pull/10, otherwise ready.

### CODE REVIEW

File: src/protocols/XDGShell.cpp [hunk 1]
```diff
@@ -7,6 +7,20 @@
 
 #define LOGM PROTO::xdgShell->protoLog
 
+void SXDGPositionerState::setAnchor(xdgPositionerAnchor anchor) {
+    this->anchor.setTop(anchor == XDG_POSITIONER_ANCHOR_TOP || anchor == XDG_POSITIONER_ANCHOR_TOP_LEFT || anchor == XDG_POSITIONER_ANCHOR_TOP_RIGHT);
```
Comment (vaxerski) [inline src/protocols/XDGShell.cpp:11]: no this->

File: src/protocols/XDGShell.cpp [hunk 2]
```diff
@@ -512,125 +526,99 @@ CXDGPositionerRules::CXDGPositionerRules(SP<CXDGPositionerResource> positioner)
...
+        if (effectiveX < constraint.x) {
+            auto diff  = constraint.x - effectiveX;
+            effectiveX = constraint.x;
+            width -= diff;
+        }
 
-        if (success)
-            return test.translate(-parentCoord - constraint.pos());
+        auto effectiveX2 = effectiveX + width;
+        if (effectiveX2 > constraint.extent().x) {
```
Comment (vaxerski) [inline src/protocols/XDGShell.cpp:603]: here and everywhere else no {} around short ifs

**Review by vaxerski — COMMENTED** [review]
cool, if it works it works

**Comment by outfoxxed** [comment]
done

File: src/protocols/XDGShell.cpp [hunk 3]
```diff
@@ -512,125 +526,93 @@ CXDGPositionerRules::CXDGPositionerRules(SP<CXDGPositionerResource> positioner)
...
+    auto width  = state.requestedSize.x;
+    auto height = state.requestedSize.y;
 
-    if (success)
-        return predictedBox.translate(-parentCoord - constraint.pos());
+    auto anchorX = state.anchor.left() ? anchorRect.x : state.anchor.right() ? anchorRect.extent().x : anchorRect.middle().x;
+    auto anchorY = state.anchor.top() ? anchorRect.y : state.anchor.bottom() ? anchorRect.extent().y : anchorRect.middle().y;
 
-    CBox test = predictedBox;
+    auto calcEffectiveX = [&]() { return state.gravity.left() ? anchorX - width : state.gravity.right() ? anchorX : anchorX - width / 2; };
```
Comment (vaxerski) [inline src/protocols/XDGShell.cpp:540]: can we have the edge padding back

Comment (outfoxxed) [reply src/protocols/XDGShell.cpp:540]: was there padding before? and how much

Comment (outfoxxed) [reply src/protocols/XDGShell.cpp:540]: ok found it

**Comment by outfoxxed** [comment]
added padding

**Review by vaxerski — DISMISSED** [review]
sankyu

**Comment by vaxerski** [comment]
wait for a bit I'll bump utils and then you bump cmakelists and meson min ver

**Comment by vaxerski** [comment]
https://github.com/hyprwm/hyprutils/releases/tag/v0.2.1

also @fufexan nix bump?

**Comment by fufexan** [comment]
Pushed.

**Comment by vaxerski** [comment]
top

**Comment by outfoxxed** [comment]
done

**Review by vaxerski — APPROVED** [review]

**Pull request merged**